book/forms.rst

```diff
@@ -621,6 +621,57 @@ large or whether you tried to submit text in a number field.
 
 .. _book-forms-type-reference:
 
+
+Groups based on a callable service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you need some advanced logic to determine the validation groups that cannot
+be simply determined by a simple callback (for example that requires a few 
+services) you can use a service. Implement a service that implements __invoke
+which accepts a FormInterface as a parameter.
+
+    use Symfony\Component\Form\FormInterface;
+    
+    class ValidationGroupResolver
+    {
+        private $service1;
+        private $sevice2;
+        public function __construct($service1,$service2)
+        {
+            $this->service1 = $service1;
+            $this->service2 = $service2;
+        }
+        
+        public function __invoke(FormInterface $form)
+        {
+            //figure out what groups to apply and return an array
+        }
+    }
+
+Then in your form, inject the resolver and set it as the validation_groups.
+
+    use Symfony\Component\OptionsResolver\OptionsResolver;
+    use Symfony\Component\Form\AbstractType
+    
+    class MyClassType extends AbstractType
+    {
+        private $groupResolver;
+        public function __construct($groupResolver)
+        {
+            $this->groupResolver = $groupResolver;
+        }
+        
+        // ...
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver->setDefaults(array(
+                'validation_groups' => $this->groupResolver,
```

Comment:
Since at this point your are invoking your group resolver through the magic `__invoke()` method, how is this different that for instance invoking an specific method of your `ValidationGroupResolver` class? something like: `$this->groupResolver->someMethod()`.